Return zero array from np.dot when the arguments are empty.

AndrewEckart opened the pull request:
When np.dot(a, b) is called with empty argments, the result can be non-empty if the zero dimension is eliminated (e.g. a is m x 0 and b is 0 x n). In these cases, we should return an array of zeros. Previously, the result was initialized with np.empty, so the it contained garbage values.

Fixes #5539

<!--

Thanks for wanting to contribute to Numba :)

First, if you need some help or want to chat to the core developers, please
visit https://gitter.im/numba/numba for real time chat or post to the Numba
mailing list https://groups.google.com/a/continuum.io/forum/#!forum/numba-users.

Here's some guidelines to help the review process go smoothly.

0. Please write a description in this text box of the changes that are being
   made.

1. Please ensure that you have written units tests for the changes made/features
   added.

2. If you are closing an issue please use one of the automatic closing words as
   noted here: https://help.github.com/articles/closing-issues-using-keywords/

3. If your pull request is not ready for review but you want to make use of the
   continuous integration testing facilities please label it with `[WIP]` and
   then remove the label when you'd like it to be reviewed.

4. Once review has taken place please do not add features or make changes out of
   the scope of those requested by the reviewer (doing this just add delays as
   already reviewed code ends up having to be re-reviewed/it is hard to tell
   what is new etc!). Further, please do not rebase your branch on master/force
   push/rewrite history, doing any of these causes the context of any comments
   made by reviewers to be lost. If conflicts occur against master they should
   be resolved by merging master into the branch used for making the pull
   request.

Many thanks in advance for your cooperation!

-->

stuartarchibald commented:
Thanks for the patch. The fix seems good. Please could you add/update appropriate unit tests in here: https://github.com/numba/numba/blob/master/numba/tests/test_linalg.py#L30 to ensure that the change is correct. Many thanks.

AndrewEckart commented:
Seems my fix is working for `dot_2`, but `dot_3` also needs to be fixed, and I'm not quite familiar enough with the lowered API to get it working. I was trying to do something along the lines of the `numpy_zeros_nd` implementation with `_zero_fill_array`.

AndrewEckart commented:
Okay, I got it working for both `dot_2` and `dot_3`. I'm very much a beginner with the IR Builder, so please let me know if there's a cleaner way to write things.

gmarkall commented:
/azp run

stuartarchibald commented:
/AzurePipelines run

stuartarchibald commented on numba/np/linalg.py:
```diff
@@ -611,8 +618,15 @@ def dot_3_vm(context, builder, sig, args):
 
-    call_xxgemv(context, builder, do_trans, mty, m_shapes, m_data,
-                v_data, out.data)
+    zero = ir.Constant(intp_t, 0)
```
There's a preference in the code base for using the `context` where possible, e.g.
`zero = context.get_constant(types.intp, 0)`
(same applies throughout).

stuartarchibald commented on numba/tests/test_linalg.py:
```diff
@@ -169,16 +168,21 @@ def samples(m, n):
-        if pyfunc3 is not None:
+
+        for m, n in [(2, 3),
+                     (3, 0)
```
What about e.g. `(0, 3)` ?

stuartarchibald commented on numba/tests/test_linalg.py:
```diff
@@ -229,10 +233,12 @@ def samples(m, n, k):
         for m, n, k in [(2, 3, 4),  # Generic matrix * matrix
                         (1, 3, 4),  # 2d vector * matrix
                         (1, 1, 4),  # 2d vector * 2d vector
+                        (3, 2, 0)   # Empty matrix * empty matrix
```
What about e.g. `(0, 0, 3)` ?

AndrewEckart commented:
> With the new use cases, I think they will probably need explicitly catching as per the current "inner dimension is zero" cases, at present, these new cases are manifesting as an error in the BLAS call, `XERBLA` is complaining!

Updated to deal with cases where one or more arguments is empty because an outer dimension is zero and added test cases to cover each case.

It's a little weird because in these cases, the output is empty, but I'm still calling `cgutils.memset` with `size=0` (which does nothing). I could add more conditional logic to avoid that if it would be preferable.

stuartarchibald commented:
> > With the new use cases, I think they will probably need explicitly catching as per the current "inner dimension is zero" cases, at present, these new cases are manifesting as an error in the BLAS call, `XERBLA` is complaining!
> 
> Updated to deal with cases where one or more arguments is empty because an outer dimension is zero and added test cases to cover each case.
> 
> It's a little weird because in these cases, the output is empty, but I'm still calling `cgutils.memset` with `size=0` (which does nothing). I could add more conditional logic to avoid that if it would be preferable.

Thanks for patching that too. Until demonstrated as a bottleneck, I think just leaving a `memset` of zero bytes is fine, it's a function call vs branch, and then obviously complexity that goes into handling it.
